jwala-services/src/main/java/com/cerner/jwala/service/group/impl/spring/component/GroupStateNotificationServiceImpl.java

```diff
@@ -66,22 +76,51 @@ public void retrieveStateAndSend(final Identifier id, final Class aClass) {
             }
-            for (final JpaGroup group: groups) {
-                final Long jvmStartedCount = jvmCrudService.getJvmStartedCount(group.getName());
-                final Long jvmStoppedCount = jvmCrudService.getJvmStoppedCount(group.getName());
-                final Long jvmForciblyStoppedCount = jvmCrudService.getJvmForciblyStoppedCount(group.getName());
-                final Long jvmCount = jvmCrudService.getJvmCount(group.getName());
-                final Long webServerStartedCount = webServerCrudService.getWebServerStartedCount(group.getName());
-                final Long webServerStoppedCount = webServerCrudService.getWebServerStoppedCount(group.getName());
-                final Long webServerCount = webServerCrudService.getWebServerCount(group.getName());
-                final CurrentState<Group, OperationalState> groupState = new CurrentState<>(new Identifier<Group>(group.getId()),
-                        null, DateTime.now(), StateType.GROUP, webServerCount, webServerStartedCount,
-                        webServerStoppedCount, jvmCount, jvmStartedCount, jvmStoppedCount, jvmForciblyStoppedCount);
+            fetchStates(groups, true);
+        }
+        LOGGER.debug("Thread locked on {} and {} released!", id, aClass);
+    }
+
+    /**
+     * Fetch the jvm and web server states for a list of groups and save them in a map
+     * @param groups the groups whose jvm and web server states are to be retrieved
+     * @param send if true, the map is sent to a topic so a client process it
+     */
+    private void fetchStates(final List<JpaGroup> groups, boolean send) {
+        for (final JpaGroup group: groups) {
+            final Long jvmStartedCount = jvmCrudService.getJvmStartedCount(group.getName());
+            final Long jvmStoppedCount = jvmCrudService.getJvmStoppedCount(group.getName());
+            final Long jvmForciblyStoppedCount = jvmCrudService.getJvmForciblyStoppedCount(group.getName());
+            final Long jvmCount = jvmCrudService.getJvmCount(group.getName());
+            final Long webServerStartedCount = webServerCrudService.getWebServerStartedCount(group.getName());
+            final Long webServerStoppedCount = webServerCrudService.getWebServerStoppedCount(group.getName());
+            final Long webServerCount = webServerCrudService.getWebServerCount(group.getName());
+            final CurrentState<Group, OperationalState> groupState = new CurrentState<>(new Identifier<>(group.getId()),
+                    null, DateTime.now(), StateType.GROUP, webServerCount, webServerStartedCount,
+                    webServerStoppedCount, jvmCount, jvmStartedCount, jvmStoppedCount, jvmForciblyStoppedCount);
+
+            if (send) {
                 messagingService.send(groupState);
-                LOGGER.debug("Group '{}' state = {}", group.getName(), groupState);
+            }
+
+            groupStateMap.put(group.getName(), groupState);
```

Comment:
Hi Arvindo, fetch state is used inside a synchronized block already. With regards to someone reading the map while writing I know there will be a race condition but I'm not overly concerned with the reading part since the said operation is measure against a point in time which can change in an instant.